### ISSUE TITLE
New Check: Alphabetical declaration order

### ISSUE THREAD

**Issue by hsellou** [sevntu-checkstyle/sevntu.checkstyle#614]
Hello,

Is it possible to have a rule checking declaration in alphabetical order ?

Maybe it's already possible with the `CustomDeclarationOrderCheckTest` check but I doubt it.

Thanks

**Comment by romani** [comment]
No, it is not possible, we need to create a new Check on it. 
It is probably reasonable to extend CustomDeclarationOrderCheck or DeclarationOrderCheck to let them make ordering in a group (fields, methods, ctors).

Please share examples of real code and what you expect it to be (2 files).

**Comment by hsellou** [comment]
Yes it would be nice to keep grouping of fields, method, ctors at the same time.
My concerns is essentially to avoid conflicts during a git merge. 

I'll try to upload a sample with all use cases asap.

Thanks !!

**Comment by romani** [comment]
> My concerns is essentially to avoid conflicts during a git merge.

I do not see how this is related, but lets see your examples.

**Comment by hsellou** [comment]
A example before/after for the new check:
[SevntuCheckstyle614.zip](https://github.com/sevntu-checkstyle/sevntu.checkstyle/files/1421337/SevntuCheckstyle614.zip)

> I do not see how this is related, but lets see your examples.

It's not directly related to this rule.
It's just that some of my colleagues use their IDEs to sort their methods (in alphabetical order) and some do not.
Git merges turn out to be complex for little.

**Comment by romani** [comment]
ok, off topic:
I do not recommend such style at all, it is better to do sorting by minimizing distances between methods - https://github.com/sevntu-checkstyle/methods-distance , one day this will be past of some Check.

**Comment by romani** [comment]
I am confident that sorting in group should be and option existing Checks and we do not need special Check for this.

@rnveach , please review and confirm your opinion on this. If you are ok. We need to change issue title and descriptin and make it approved.

**Comment by rnveach** [comment]
@romani I am fine with an option in one of the checks you mentioned for doing this,
but if all @hsellou wants is to have the fields sorted by name and no other modifiers or such, then I don't think `DeclarationOrderCheck` would work for him. So it would have to be `CustomDeclarationOrderCheckTest` but how will you define if all fields/methods should be sorted first or based on another criteria before the name like a modifier.

Simply: How will you define `Order By Name, Modifier` versus `Order By Modifier, Name`?
His example looks like he wants `Order By Modifier, Name`.

**Comment by romani** [comment]
@rnveach, Example of custom order - https://github.com/sevntu-checkstyle/sevntu.checkstyle/blob/master/sevntu-checks/sevntu-checks.xml#L45 , Each group like `Field(protected)` will be sorted alphabetically if user wants this.

@hsellou , why getter and setters are not sorted ? Do you still want them to be coupled close to each other?

**Comment by hsellou** [comment]
@romani 

Yeah, Intellij groups getters and setters by attribute.
But I can accept a solution in which we sort getters and then setters like this:

```java
public class SevntuCheckstyle614After {

  public String fieldG;
  public String fieldH;

  private String fieldE;
  private String fieldF;

  public SevntuCheckstyle614After(String fieldE, String fieldF) {
    this.fieldE = fieldE;
    this.fieldF = fieldF;
  }

  public String getFieldE() {
    return fieldE;
  }
  
  public String getFieldF() {
    return fieldF;
  }

  public void setFieldE(String fieldE) {
    this.fieldE = fieldE;
  }

  public void setFieldF(String fieldF) {
    this.fieldF = fieldF;
  }

  public void doSometingC() { }

  public void doSometingD() { }

  private void doSomethingA() {}

  private void doSomethingB() {}
}
```

**Comment by romani** [comment]
@rnveach , are you ok now ?

**Comment by romani** [comment]
@hsellou , I changed issue title and approved issue. Please be welcome to send update to Check to get what you want.
